working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+# Horizontal inference
+
+Author: paulberry@google.com, Version: 1.0 (See [Changelog](#Changelog) at end)
+
+Horizontal inference allows certain arguments of an invocation (specifically
+function literals) to be type inferred in a context derived from the type of
+other arguments.
+
+## Motivation
+
+The key motivating example is the Iterable.fold method, which has the following
+signature:
+
+```dart
+abstract class Iterable<E> {
+  T fold<T>(T initialValue, T combine(T previousValue, E element)) { ... }
+}
+```
+
+A typical usage of `fold` today looks like this:
+
+```dart
+Iterable<int> values = ...;
+int largestValue = values.fold(0, (a, b) => a < b ? b : a);
+```
+
+Note the use of the explicit type `int` for `largestValue`.  We would like for
+the user to be able to drop this explicit type, and instead write:
+
+```dart
+Iterable<int> values = ...;
+var largestValue = values.fold(0, (a, b) => a < b ? b : a);
+```
+
+Today this doesn't work, because without the leading `int`, the downwards
+inference phase of type inference has no information with which to choose a
+preliminary type for the type parameter `T`, so the type context used for
+inference of the function literal `(a, b) => a < b ? b : a` is `? Function(?,
+int)`.  Hence, `a` gets assigned a static type of `Object?`, and this leads to a
+compile error at `a < b`.
+
+With horizontal inference, type inference first visits the `0`, and uses its
+static type to assign a preliminary type of `int` to the type parameter `T`.
+Then, when it proceeds to perform type inference on the function literal, the
+type context is `int Function(int, int)`.  This results in `a` being assigned a
+static type of `int`, which is what the user intends.
+
+## Type inference algorithm for invocations
+
+The full algorithm for type inference of invocations is expanded as follows.
```

Comment:
I see where you're coming from, however I prefer keeping the two improvements in the same proposal, for two reasons. 
 (1) Both improvements have already been implemented in the analyzer, and a full implementation in the front end should be finished soon, so separating the two proposals wouldn't actually give us any benefit in terms of implementation flexibility.  (2) Since the algorithm for type inference of invocations (as it exists in Dart 2.18) is not currently specified anywhere, I believe that breaking the proposal into two wouldn't actually be any easier for readers to understand, because although the two individual proposals would be shorter, they would both be written as deltas against an implicit and unwritten base algorithm.

Reply:
OK!